File: Sources/RedlineCore/Sidecar.swift
// The usage sidecar, written as well as read.
//
// Three separate projects (ClaudeHUD, claude-monitor, CodexBar) converged on the same idea:
// one small JSON file holding the current rate-limit windows, written by whoever has them and
// read by whoever needs them. RedLine already reads that shape. This publishes it too, so a
// status line, a tmux strip, or another monitor can take RedLine's reading instead of every
// tool racing for the same source.
//
// Standard keys first, in the shape the other tools already parse. Anything RedLine-specific
// lives under a "redline" object that a foreign reader can ignore without losing the windows.
import Foundation

public enum Sidecar {
    /// Where RedLine publishes. Deliberately not the file the statusline feeder writes:
    /// reading our own output back would be a loop, and the feeder's file is Claude Code's
    /// word rather than ours.
    public static func publishURL(home: URL? = nil) -> URL {
        return AppPaths.data("usage-snapshot.json", in: home)
    }

    /// How old an external sidecar may be before it is ignored. Same tolerance RedLine
    /// applies to its own feed, since it is the same kind of file.
    public static let defaultFreshness: TimeInterval = 900

    public struct Totals {
        public let io: Int
        public let cost: Double
        public let priced: Bool

        public init(io: Int, cost: Double, priced: Bool) {
            self.io = io
            self.cost = cost
            self.priced = priced
        }
    }

    /// Builds the payload. Kept separate from writing so the shape can be asserted in tests
    /// without touching the filesystem.
    public static func payload(windows: [LimitWindow], updatedAt: Date, producer: String,
                               today: Totals? = nil, week: Totals? = nil,
                               limitsAsOf: Date? = nil) -> [String: Any] {
        let iso = ISO8601DateFormatter()
        var out: [String: Any] = [
            "updated_at": iso.string(from: updatedAt),
            "source": "redline",
            "producer": producer,
        ]

        let claude = windows.filter {
            $0.provider.caseInsensitiveCompare("Claude") == .orderedSame
        }
        for w in claude where w.key == "five_hour" || w.key == "seven_day" {
            var block: [String: Any] = [
                // Both spellings on purpose: the statusline payload says used_percentage,
                // the usage endpoint says utilization, and readers exist for each.
                "used_percentage": w.utilization,
                "utilization": w.utilization,
            ]
            if let r = w.resetsAt { block["resets_at"] = iso.string(from: r) }
            out[w.key] = block
        }

        let scoped = claude.filter { $0.key.hasPrefix("seven_day_") }
        if !scoped.isEmpty {
            out["model_scoped"] = scoped.map { w -> [String: Any] in
                var entry: [String: Any] = [
                    "display_name": displayName(forScopedKey: w.key),
                    "utilization": w.utilization,
                ]
                if let r = w.resetsAt { entry["resets_at"] = iso.string(from: r) }
                return entry
            }
        }

        var extra: [String: Any] = [
            "windows": windows.filter { !$0.isUninformative }.map { w -> [String: Any] in
                var entry: [String: Any] = [
                    "provider": w.provider,
                    "key": w.key,
                    "utilization": w.utilization,
                    "source": w.source.rawValue,
                ]
                if let r = w.resetsAt { entry["resets_at"] = iso.string(from: r) }
                return entry
            },
        ]
        if let limitsAsOf { extra["claude_limits_as_of"] = iso.string(from: limitsAsOf) }
        if let today { extra["today"] = totalsBlock(today) }
        if let week { extra["week"] = totalsBlock(week) }
        out["redline"] = extra
        return out
    }

    private static func totalsBlock(_ t: Totals) -> [String: Any] {
        [
            "tokens": t.io,
            "cost_usd": t.cost,
            // Tokens are counted by the provider; the dollar figure is arithmetic over a
            // pricing table, and an unpriced model means even that is incomplete.
            "tokens_basis": Provenance.official.rawValue,
            "cost_basis": Provenance.localEstimate.rawValue,
            "cost_partial": !t.priced,
        ]
    }

    /// "seven_day_opus" -> "Opus". The key was derived from the display name in the first
    /// place, so this puts back what the round trip took out.
    static func displayName(forScopedKey key: String) -> String {
        let slug = key.hasPrefix("seven_day_") ? String(key.dropFirst(10)) : key
        return slug.split(separator: "_")
            .map { $0.prefix(1).uppercased() + $0.dropFirst() }
            .joined(separator: " ")
    }

    @discardableResult
    public static func publish(windows: [LimitWindow], updatedAt: Date = Date(),
                               producer: String, today: Totals? = nil, week: Totals? = nil,
                               limitsAsOf: Date? = nil, to url: URL? = nil) -> Bool {
        let url = url ?? publishURL()
        let json = payload(windows: windows, updatedAt: updatedAt, producer: producer,
                           today: today, week: week, limitsAsOf: limitsAsOf)
        guard let data = try? JSONSerialization.data(withJSONObject: json,
                                                     options: [.prettyPrinted, .sortedKeys])
        else {
            Diag.log.error("sidecar.encode_failed", "could not encode sidecar payload")
            return false
        }
        do {
            try FileManager.default.createDirectory(at: url.deletingLastPathComponent(),
                                                    withIntermediateDirectories: true)
            try data.write(to: url, options: .atomic)
            try? FileManager.default.setAttributes([.posixPermissions: 0o600],
                                                   ofItemAtPath: url.path)
            return true
        } catch {
            Diag.log.error("sidecar.write_failed", "could not publish sidecar",
                           ["path": url.path, "error": String(describing: error)])
            return false
        }
    }

    public static func remove(at url: URL? = nil) {
        try? FileManager.default.removeItem(at: url ?? publishURL())
    }

    /// Reads someone else's sidecar, when the user has pointed at one. Same parser as the
    /// feed, since it is the same contract; the extra rules here are about trusting a file
    /// this app did not write: absolute path, .json, and fresh enough to be about now.
    public static func readExternal(path: String, freshness: TimeInterval = defaultFreshness,
                                    now: Date = Date()) -> StatuslineSnapshot? {
        guard let url = validExternalPath(path) else { return nil }
        guard let snap = StatuslineFeed.read(path: url, now: now) else { return nil }
        guard let at = snap.updatedAt, now.timeIntervalSince(at) <= freshness,
              !snap.isEmpty else { return nil }
        return snap
    }

    /// Nil for anything that is not an absolute path to a .json file. A relative path would
    /// resolve against whatever directory the process happened to start in.
    public static func validExternalPath(_ path: String) -> URL? {
        let trimmed = path.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        let expanded = (trimmed as NSString).expandingTildeInPath
        guard expanded.hasPrefix("/"), expanded.lowercased().hasSuffix(".json") else {
            return nil
        }
        return URL(fileURLWithPath: expanded)
    }
}

File: windows/src/Redline.Core/Sidecar.cs
// The usage sidecar, published in the shape ClaudeHUD, claude-monitor and CodexBar already read.
// Standard keys first; anything RedLine-specific lives under "redline", which others can ignore.
using System.Text.Json.Nodes;

namespace Redline.Core;

public static class Sidecar
{
    /// Not the feeder's file: reading our own output back would be a loop.
    public static string PublishUrl(string? home = null) =>
        RedlineHome.Join(home ?? RedlineHome.Url, ".local/share/redline/usage-snapshot.json");

    /// Seconds an external sidecar may age before it is ignored, same as RedLine's own feed.
    public const double DefaultFreshness = 900;

    public sealed record Totals(int Io, double Cost, bool Priced);

    internal static string Iso(DateTimeOffset d) => DiagnosticsLog.Stamp(d);

    /// Builds the payload apart from writing it, so the shape is testable without a filesystem.
    public static JsonObject Payload(IEnumerable<LimitWindow> windows, DateTimeOffset updatedAt,
                                     string producer, Totals? today = null, Totals? week = null,
                                     DateTimeOffset? limitsAsOf = null)
    {
        var all = windows.ToList();
        var output = new JsonObject
        {
            ["updated_at"] = Iso(updatedAt),
            ["source"] = "redline",
            ["producer"] = producer,
        };

        var claude = all.Where(w => string.Equals(w.Provider, "Claude", StringComparison.OrdinalIgnoreCase)).ToList();
        foreach (var w in claude.Where(w => w.Key is "five_hour" or "seven_day"))
        {
            // Both spellings on purpose: readers exist for the statusline's and the endpoint's
            var block = new JsonObject { ["used_percentage"] = w.Utilization, ["utilization"] = w.Utilization };
            if (w.ResetsAt is { } r) block["resets_at"] = Iso(r);
            output[w.Key] = block;
        }

        var scoped = claude.Where(w => w.Key.StartsWith("seven_day_", StringComparison.Ordinal)).ToList();
        if (scoped.Count > 0)
        {
            output["model_scoped"] = new JsonArray(scoped.Select(w =>
            {
                var entry = new JsonObject
                {
                    ["display_name"] = DisplayName(w.Key),
                    ["utilization"] = w.Utilization,
                };
                if (w.ResetsAt is { } r) entry["resets_at"] = Iso(r);
                return (JsonNode)entry;
            }).ToArray());
        }

        var extra = new JsonObject
        {
            ["windows"] = new JsonArray(all.Where(w => !w.IsUninformative).Select(w =>
            {
                var entry = new JsonObject
                {
                    ["provider"] = w.Provider,
                    ["key"] = w.Key,
                    ["utilization"] = w.Utilization,
                    ["source"] = w.Source.RawValue(),
                };
                if (w.ResetsAt is { } r) entry["resets_at"] = Iso(r);
                return (JsonNode)entry;
            }).ToArray()),
        };
        if (limitsAsOf is { } asOf) extra["claude_limits_as_of"] = Iso(asOf);
        if (today is not null) extra["today"] = TotalsBlock(today);
        if (week is not null) extra["week"] = TotalsBlock(week);
        output["redline"] = extra;
        return output;
    }

    private static JsonObject TotalsBlock(Totals t) => new()
    {
        ["tokens"] = t.Io,
        ["cost_usd"] = t.Cost,
        // Tokens are the provider's count; dollars are arithmetic over a pricing table
        ["tokens_basis"] = Provenance.Official.RawValue(),
        ["cost_basis"] = Provenance.LocalEstimate.RawValue(),
        ["cost_partial"] = !t.Priced,
    };

    /// "seven_day_opus" -> "Opus": puts back what the display-name round trip took out.
    internal static string DisplayName(string scopedKey)
    {
        var slug = scopedKey.StartsWith("seven_day_", StringComparison.Ordinal) ? scopedKey[10..] : scopedKey;
        return string.Join(" ", slug.Split('_', StringSplitOptions.RemoveEmptyEntries)
            .Select(p => p[..1].ToUpperInvariant() + p[1..]));
    }

    public static bool Publish(IEnumerable<LimitWindow> windows, string producer,
                               DateTimeOffset? updatedAt = null, Totals? today = null,
                               Totals? week = null, DateTimeOffset? limitsAsOf = null, string? to = null)
    {
        var path = to ?? PublishUrl();
        var json = Payload(windows, updatedAt ?? DateTimeOffset.UtcNow, producer, today, week, limitsAsOf);
        string text;
        try { text = SortKeys(json)!.ToJsonString(Json.Pretty); }
        catch
        {
            Diag.Log.Error("sidecar.encode_failed", "could not encode sidecar payload");
            return false;
        }
        try
        {
            Json.WriteAtomic(path, text);
            return true;
        }
        catch (Exception e)
        {
            Diag.Log.Error("sidecar.write_failed", "could not publish sidecar",
                new() { ["path"] = path, ["error"] = e.Message });
            return false;
        }
    }

    // Sorted keys, like JSONSerialization's .sortedKeys, so the file diffs cleanly
    internal static JsonNode? SortKeys(JsonNode? node) => node switch
    {
        JsonObject o => new JsonObject(o.OrderBy(kv => kv.Key, StringComparer.Ordinal)
            .Select(kv => KeyValuePair.Create(kv.Key, SortKeys(kv.Value)))),
        JsonArray a => new JsonArray(a.Select(SortKeys).ToArray()),
        _ => node?.DeepClone(),
    };

    public static void Remove(string? at = null)
    {
        try { File.Delete(at ?? PublishUrl()); } catch { }
    }

    /// Someone else's sidecar: absolute .json path, fresh enough to be about now, not empty.
    public static StatuslineSnapshot? ReadExternal(string path, double freshness = DefaultFreshness,
                                                   DateTimeOffset? now = null)
    {
        var t = now ?? DateTimeOffset.UtcNow;
        if (ValidExternalPath(path) is not { } p) return null;
        if (StatuslineFeed.Read(p, t) is not { } snap) return null;
        if (snap.UpdatedAt is not { } at || (t - at).TotalSeconds > freshness || snap.IsEmpty) return null;
        return snap;
    }

    /// Null for anything but an absolute path to a .json file; "~" expands to the account home.
    public static string? ValidExternalPath(string path)
    {
        var trimmed = path.Trim();
        if (trimmed.Length == 0) return null;
        var expanded = trimmed == "~" ? RedlineHome.AccountHome
            : trimmed.StartsWith("~/") || trimmed.StartsWith("~\\")
                ? Path.Combine(RedlineHome.AccountHome, trimmed[2..]) : trimmed;
        if (!Path.IsPathFullyQualified(expanded)) return null;
        return expanded.EndsWith(".json", StringComparison.OrdinalIgnoreCase) ? expanded : null;
    }
}
